### DISCUSSION TITLE
Is libvips capable of importing and exporting HEIC images? (In addition to HEIF images)

### DISCUSSION THREAD

**Discussion by AttilaTheFun** [libvips/libvips#2627] [General]
My understanding is that HEIC images are basically a container that can store one or more HEIF images bundled together with other media such as videos. Apple uses the HEIC file format for their Live Photos feature where a HEIF image is combined with a short video clip leading up to the capture of that photo. 

I understand that full support for processing Live Photos would be out of scope, but iOS devices capture images in the HEIC format by default, even if the user is not using the Live Photo feature. I've tried passing a HEIC image into libvips and it seemed to work, but on the export side I only see options for exporting HEIF images. 

I'm worried that some iOS code is explicitly written to look for HEIC files and it will break if presented with a HEIF file instead. Is there any difference between a HEIF image exported with libvips and a HEIC image produced by iOS beyond the file extension? 

If so, would it be possible to add functionality to libvips to allow it to export HEIC images that wrap the HEIF images currently produced by libvips? It would be fine by me if the video either "fell off" of Live Photos or passed through unmodified.

**Comment by lovell** [comment]
Are you able to provide examples using the libheif command line `heif-enc` tool, e.g. something that works as expected and something that doesn't, to help demonstrate your question?

**Comment by jcupitt** [comment]
You're almost right -- HEIF is the container. A HEIC image is a HEIF container plus an image encoded with h265. libheif will need both an h265 decode library and an h265 encode library, so it's a bit tricky to set up. You'll also need to be very careful around patent issues, since especially software h265 encode may be legally difficult where you live, or where your server is hosted.

Anyway, it should all work. As Lovell says, if you've found a problem, please give a specific example to illustrate it.

**Comment by AttilaTheFun** [comment]
@lovell @jcupitt Thanks for the responses! 

While I don't have a specific case where an iOS component fails with a HEIF image but works with a HEIC image, my concern is that some places may hardcode checks for the "heic" file extension or the associated Unified Type Identifier, "public.heic". 

Once place where I suspect this may happen is with the NSActivity system that is used by the share sheet, drag-and-drop, handoff and other features. If I offer a HEIF image, whether or not Apple has handled it correctly on their end, if a 3rd party app developer only registers to handle "public.heic" UTI, the user won't be able to use the feature with their app. The way that they *should* handle this is by registering to handle their shared parent UTI, "public.heif-standard", but most developers don't do that.

Ideally I would just like to be able to encode images to HEIC files (not HEIF) for maximum compatibility on Apple devices. (Also Apple has some fancy hardware H.265 decoders that make them incredibly fast and efficient to encode and decode.)

Regarding the h265 encoder / decoder, I'm developing this for use in a docker container running Alpine Linux. The Alpine Linux package manager has a distribution of [libheif](https://pkgs.alpinelinux.org/packages?name=libheif&branch=edge) which uses [libde265](https://www.libde265.org), an open source H.265 encoder / decoder with an LGPL license.

Do you think that would be sufficient for encoding HEIC images? And if so, how would I go about encoding images to HEIC instead of HEIF?

**Reply by jcupitt** [reply]
There's no such thing as a HEIF image, I'm not sure what you mean. Maybe you're thinking of AVIF?

libde265 is decode only (hence the name). You'll need libx265 for encode, which is GPL and has a range of patent issues you need to think about.

If you set it up right, it should all work, eg. I can do:

```
vips copy x.heic y.heic
```

**Reply by AttilaTheFun** [reply]
Sorry if I was unclear, I understand that HEIF is a container that can hold images encoded in a variety of different formats including images encoded with x265 HEVC.

When I said "HEIF Images" I was referring to HEIF image containers with the ".heif" file extension as opposed to "HEIC Images", i.e. HEIF image containers with the ".heic" extension. heifload.c appears to allow for both ".heif" and ".heic" file extensions:
https://github.com/libvips/libvips/blob/cfa7f03278c3aa10bc9232ac14378817046c16c0/libvips/foreign/heifload.c#L81

One of my questions was whether there is a structural difference between HEIF image containers with the ".heif" and ".heic" file extensions. Are they interchangeable like ".jpg" and ".jpeg"? Or is ".heic" supposed to be reserved for HEIF containers with an image encoded with x265 HEVC?

For example, my Canon R5 can produce ".heif" files which I believe are using x265 because the camera can also record video in HEVC with a hardware encoder. My iPhone 13 Pro produces ".heic" files which I think are also using Apple's own hardware encoders. Are these files structurally identical beyond the file extension?

Re: the x265 encoder, you're right. That one is decode only. But Alpine also has an x265 package with the VideoLAN (VLC) encoder:
https://pkgs.alpinelinux.org/packages?name=x265&branch=edge
https://www.videolan.org/developers/x265.html

Could I use that? 

Re: Patent restrictions, my understanding is that GPLv2 just says anyone with the binary can request the source, but if I'm just using the binary on my server and not giving it away it should be okay, right?

**Reply by jcupitt** [reply]
Yes, it's just the file extension, as far as I know.

Yes, you can use libx265, but it's covered by patents, so there are legal issues to consider:

https://en.wikipedia.org/wiki/High_Efficiency_Video_Coding#Provision_for_costless_software

**Reply by AttilaTheFun** [reply]
Actually after a little research I found that the ".heic" file extension *is* semantically meaningful. 

While the container structure is the same for ".heif" and ".heic" images, ".heic" images are specifically supposed to signify that the HEIF container has an image encoded with HEVC: https://nokiatech.github.io/heif/technical.html
".heif" images are just generic HEIF containers that use some other non-HEVC coding format.

So, to back my original question, is libvips currently capable of producing HEVC-encoded HEIF files? If so, how could I configure it to do this? (E.g. using the libx265 library we discussed, patent issues notwithstanding.) 

If not, what encoding does libvips provide out of the box? And if it's not using a HEVC encoder, could I assume that those files should be given the ".heif" file extension because they do not use HEVC for encoding? Thanks!

**Reply by jcupitt** [reply]
Right, that's what I said at the beginning, .heic is a HEIF container with a h265 codec.

As I said, set up your libheif with x265 and libvips will write .heic images with eg. `vips copy x.heic y.heic`.

It also supports the AV1 compressor and will write .avif (if you've set up your libheif for that), but h265 is the default.

**Reply by AttilaTheFun** [reply]
Ah I must have misunderstood, sorry about that!
Thanks, I'll give that a shot!
